### PR TITLE
ref(compactSelect): Memoize props to improve performance

### PR DESCRIPTION
**Pull request by vuluongj20**
Memoize `MenuListItem` and the props it receives (`optionProps`, `rowProps,…)` from `CompactSelect`. This greatly reduces the number of rerenders, improving scroll performance, especially in grid mode.

**Before ——** scrolling is slow in grid mode

https://user-images.githubusercontent.com/44172267/222227655-8f68ed75-dcd4-4cd0-8726-390e16da4418.mov

---

**After ——**

https://user-images.githubusercontent.com/44172267/222227710-d3aa810f-cd66-421a-9f10-21ba58facd60.mov

### CODE REVIEW

File: static/app/components/compactSelect/listBox/option.tsx [hunk 1]
```diff
@@ -42,35 +42,52 @@ export function ListBoxOption({item, listState, size}: ListBoxOptionProps) {
     ref
   );
 
-  const checkboxSize = size === 'xs' ? 'xs' : 'sm';
+  const optionPropsMemo = useMemo(
```
Comment (malwilley) [inline static/app/components/compactSelect/listBox/option.tsx:45]: Is this necessary to memoize? It gets spread onto MenuListItem so I'd think that the object reference changing wouldn't matter

File: static/app/components/compactSelect/gridList/option.tsx [hunk 2]
```diff
@@ -64,37 +64,60 @@ export function GridListOption({node, listState, size}: GridListOptionProps) {
   const [isFocusWithin, setFocusWithin] = useState(false);
   const {focusWithinProps} = useFocusWithin({onFocusWithinChange: setFocusWithin});
 
-  const checkboxSize = size === 'xs' ? 'xs' : 'sm';
+  const rowPropsMemo = useMemo(
```
Comment (malwilley) [inline static/app/components/compactSelect/gridList/option.tsx:67]: Same question here about this being necessary

File: static/app/components/compactSelect/listBox/option.tsx (hunk 1 above)
Comment (vuluongj20) [reply static/app/components/compactSelect/listBox/option.tsx:45]: I think so! `optionProps` contains a bunch of event callbacks (`onPointerDown`, `onKeyDown`,…), each of which gets updated with a different reference every time the selection/focus state changes, regardless of whether the current option is affected by that change. So, without memoizing `optionProps`, every time we hover on a new option, all `MenuListItem`s (even the version wrapped inside `memo()`) in the list box will rerender.

Comment (vuluongj20) [reply static/app/components/compactSelect/listBox/option.tsx:45]: This wouldn't be necessary if `react-aria`'s `useOption` returned memoized props/callbacks, but it doesn't sadly.

**Review by malwilley — APPROVED** [review]

**Review by evanpurkhiser — APPROVED** [review]

**Pull request merged**